### PR TITLE
Add Bluetooth Smart API. Fixes to OCF and Board APIs.

### PR DESCRIPTION
**Pull request by zolkis**
Add Bluetooth Smart API
OCF: add support for resource translation.
Change OCF request response: use a respond() and error() method on the request object, like with the Bluetooth Smart API.
Add tables of content. Editorial changes. Fix permission steps.

### CODE REVIEW

**Comment by zolkis** [comment]
@poussa @01org/zephyr-js-developers @gabrielschulhof  @kenchris PTAL

File: api/ble/README.md [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+- [The `BluetoothCharacteristic` interface](#characteristic)
+  * `BluetoothCharacteristic` events
+    - [read](#onread)
+    - [write](#onwrite)
+    - [subscribe](#onsubscribe)
+    - [unsubscribe](#onunsubscribe)
+    - [error](#oncherror)
+  * `BluetoothCharacteristic methods
+    - [startNotifications()](#startnotifications)
+    - [startNotifications()](#stopnotifications)
```
Comment (kenchris) [inline api/ble/README.md]: should be stop

File: api/ble/README.md [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+- [The `BluetoothCharacteristicRequest` interface](#characteristicrequest)
+  * `BluetoothCharacteristicRequest` methods
+    - [respond(data)](#respond)
```
Comment (kenchris) [inline api/ble/README.md]: me wants respondWith :)

File: api/ble/README.md [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+    - [error(error)](#errormethod)
+
+Introduction
+------------
+This API exposes Bluetooth Smart functionality for Bluetooth Smart Peripheral (BSP) mode. The terminology and context is taken from the [Bluetooth 4.2. specification](https://www.bluetooth.org/DocMan/handlers/DownloadDoc.ashx?doc_id=286439).
+
+A Bluetooth Smart Peripheral (BSP) device supports the following functionality:
+- Host a [GATT (Generic Attribute Profile)](https://www.bluetooth.com/specifications/gatt) server that exposes a hierarchy of [`Service`](#service), [`Characteristic`](#characteristic) and [`Descriptor`](#descriptor) objects.
+- Broadcast (advertise) information (e.g. services) using advertisment packets. Advertisement or Scan Response data packet is a 31 byte payload used to advertise the device's capabilities and connection parameters.
+- Accept connections from Bluetooth Smart devices in Central mode that can access its services exposed through the GAP, the [Generic Access Profile](https://www.bluetooth.org/en-us/specification/assigned-numbers/generic-access-profile). A device has only one
```
Comment (kenchris) [inline api/ble/README.md]: link to central mode?

Comment (kenchris) [inline api/ble/README.md]: the GAP, the Generic Access Profile... maybe add in ( instead

File: api/ble/README.md [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+instance of the GAP service in the GATT server. The GAP service is a GATT
+based service with the service UUID as «GAP Service» defined in the
+[Bluetooth Assigned Numbers document](https://www.bluetooth.com/specifications/assigned-numbers).
+
+The [W3C Web Bluetooth](https://webbluetoothcg.github.io/web-bluetooth) API exposes functionality for Bluetooth Smart clients (Central mode).
+
+### Use cases
+
+This API is used by applications that implement a Bluetooth Peripheral device and expose sensorial data via Bluetooth.
```
Comment (kenchris) [inline api/ble/README.md]: sensorial?

File: api/ble/README.md [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+- Applications define the Bluetooth Smart services, characteristics and descriptors, then registers them with the implementation.
```
Comment (kenchris) [inline api/ble/README.md]: can define?

File: api/ble/README.md [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+- Applications can build Bluetooth advertisement packets, then start (and stop) advertising.
+- For the defined characteristics, applications should register event listeners for handling requests coming from clients. The event listeners receive the request, and should reply to them with appropriate data, or error code.
+
+<a name="apiobject">
+The API object
+--------------
+The API entry point is a [`BluetoothPeripheralDevice`](./#bpd) object that is exposed in a platform specific manner. As an example, on Node.js it can be obtained by requiring the package that implements this API. On other platforms, it can be constructed.
+
+```javascript
+var bpd = require('bluetooth-peripheral');
```
Comment (kenchris) [inline api/ble/README.md]: what does the d stand for in bpt? maybe just use something descriptable like ```var peripheral```

File: api/ble/README.md [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+| addressType | String enum | no | `undefined` | Bluetooth Device Address Type |
+| name      | String | yes | `undefined` | the Bluetooth Device Name in scan responses |
+| enabled   | boolean | no | false | whether Bluetooth and this device is enabled |
+| services  | array of Service objects | no | [] | Primary services on this device |
+
+<a name="bd_addr">
+The `address` property is a 48 bit value (BD_ADDR) identifying a Bluetooth Smart device. On the UI level the Bluetooth address MUST be represented as 12 hexadecimal characters, possibly divided into sub-parts separated by ‘:’ (e.g., ‘000C3E3A4B69’ or ‘00:0C:3E:3A:4B:69’). On the UI level any number MUST have the most significant bit on left ordering.
+
+<a name="bd_addr_type">
+The `addressType` property can take the following values: `"public"`, `"static-random"`, `"private-resolvable"`, `"private"`.
```
Comment (kenchris) [inline api/ble/README.md]: , and "private"

File: api/ble/README.md [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+
+<a name="bd_name">
+The `name` property represents the Bluetooth device name that a Bluetooth device
+exposes to remote devices. It and can be up to 248 bytes encoded in UTF-8, therefore on the UI level it may be restricted to as few as 62 characters if codepoints outside the range U+0000 to U+007F are used. A device cannot expect that a general remote device is able to handle more than the first 40 characters of the Bluetooth device name. If a remote device has limited display capabilities, it may use only the first 20 characters.
```
Comment (kenchris) [inline api/ble/README.md]: It and can? sounds weird

Comment (kenchris) [inline api/ble/README.md]: encoded *as* UTF-8

File: api/ble/README.md [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+
+<a name="bd_enabled">
+The `enabled` property MUST be set to `true` if the Bluetooth Smart and the device is enabled, and otherwise to `false`.
```
Comment (kenchris) [inline api/ble/README.md]: is enabled *on the device* ?

File: api/ble/README.md [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+
+<a name="bd_services">
+The `services` property is a read-only array (sequence) of [`BluetoothService`](#service) objects that represents the primary GATT services on this device. A primary service is visible at device root level, and it may contain other services.
```
Comment (kenchris) [inline api/ble/README.md]: on the peripheral device (instead of this)

Comment (kenchris) [inline api/ble/README.md]: what is device root level?

File: api/ble/README.md [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+
+#### `BluetoothPeripheralDevice` events
+
+| Event name        | Event callback argument |
+| --------------    | ----------------------- |
+| *enabledchange*   | N/A                     |
+| *connect*         | String (client Bluetooth Device Address) |
```
Comment (kenchris) [inline api/ble/README.md]: which is Device and Address upper case but client not

File: api/ble/README.md [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+<a name="ondisconnect"></a>
+The `disconnect` event is emitted when a connected client device is disconnected. Listeners are invoked with the client's [Bluetooth Device Address](#bd_addr) as a string argument.
+
+<a name="onerror"></a>
+The `error` event is emitted when a Bluetooth error needs to be reported to the application. Listeners are invoked with an [`Error`](https://nodejs.org/api/errors.html#errors_class_error) object as argument.
+
+#### `BluetoothPeripheralDevice` methods
+
+<a name="enable"></a>
+##### The `enable()` method
```
Comment (kenchris) [inline api/ble/README.md]: You say "request" but don't talk about what happens if it fails... does it throw? reject promiose?

File: api/ble/README.md [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+###### The `advertisement` argument
+Advertisement packet data can be of the following types:  string (e.g. flags), device name, transmit power, service UUID (16, 32, or 128 bit), service data, public or random target address, advertising interval, device address, Bluetooth Smart role, URI, or manufacturer data. Applications are expected to build the binary advertisement packets based on the Bluetooth specifications as a `Buffer` object. This API provides helpers for the most common data types. It is an object with properties described in the `startadvertising()` method steps.
+
+The `startadvertising()` method runs the following steps:
+- Return a [`Promise`](../README.md/#promise) object `promise` and continue [in parallel](https://html.spec.whatwg.org/#in-parallel).
+- If `options.connectable` is not `true`, then set `options.connectable` to `false`.
+- If `options.minInterval` is `undefined`, set `options.minInterval` to a platform default value. Otherwise, if it is a number, and it is smaller than 100 or bigger than 32000, reject `promise` with `TypeError`. Otherwise set `options.minInterval` to that value rounded down to the nearest value dividable by 100.
```
Comment (kenchris) [inline api/ble/README.md]: , after Otherwise

File: api/ble/README.md [hunk 14]
```diff
@@ -0,0 +1,305 @@
...
+- If `advertisement.data` is a [`Buffer`](../README.md/#buffer) object, let `buffer` take that value.
+- Otherwise let `buffer` be an empty `Buffer`, and use the properties of `advertisement` (when defined) to prepare `buffer` based on the following properties of `advertisement`. The algorithm on how to prepare `buffer` is described in the Bluetooth specification, and the platform may have an API to support that.
+  * `uuids` is an array of Bluetooth UUID strings representing services; if specified, implementations SHOULD add them to the advertisement packet `buffer` compliant to the Bluetooth specification.
+  * `serviceData` is an object with two properties, a UUID string `uuid` and a `Buffer` object `data`. Implementations should add `uuid` and `data` to `buffer`, if there is sufficient space in the advertisement packet. If there is no space, skip to the next step.
+  * `manufacturerData` is an object with two properties, a number `manufacturerId` and a `Buffer` data. Implementations should add `manufacturerId` and `data` to `buffer`, if there is sufficient space in the advertisement packet. If there is no space, skip to the next step.
```
Comment (kenchris) [inline api/ble/README.md]: if there is not sufficient space (instead of no space)

File: api/ble/README.md (hunk 2 above)
Comment (zolkis) [inline api/ble/README.md]: data is optional, and it would look weird calling `respondWith()` with no args.

File: api/ble/README.md (hunk 3 above)
Comment (zolkis) [inline api/ble/README.md]: no direct link to Central mode, since it's in the Core spec pdf

File: api/ble/README.md (hunk 4 above)
Comment (zolkis) [inline api/ble/README.md]: changed to sensor data

File: api/ble/README.md [hunk 15]
```diff
@@ -0,0 +1,305 @@
...
+
+<a name="apiobject">
+The API object
+--------------
+The API entry point is a [`BluetoothPeripheralDevice`](./#bpd) object that is exposed in a platform specific manner. As an example, on Node.js it can be obtained by requiring the package that implements this API. On other platforms, it can be constructed.
+
+```javascript
+var bpd = require('bluetooth-peripheral');
+```
+If the functionality is not supported by the platform, `require` should throw `NotSupportedError`. If there is no permission for using the functionality, `require` should throw `SecurityError`.
```
Comment (zolkis) [inline api/ble/README.md]: ok, changed the text to "If the underlying platform doesn't allow using the functionality, `require` should throw `SecurityError`."

File: api/ble/README.md (hunk 11 above)
Comment (zolkis) [inline api/ble/README.md]: "Bluetooth Device Address" is a name in the core spec. Client is common noun.

File: api/ble/README.md (hunk 12 above)
Comment (zolkis) [inline api/ble/README.md]: It is concise. It says 
> When the device is enabled, implementations should set the `enabled` property to `true` and emit the `enabledchange` event.

That's it.

**Comment by zolkis** [comment]
Fixed most comments.

File: api/ble/README.md (hunk 11 above)
Comment (kenchris) [inline api/ble/README.md]: still it looks wrong, why cant client be upper case? of the Bluetooth Device Address of the Client

Comment (zolkis) [inline api/ble/README.md]: Maybe looks wrong, but it's correct. If Client would be there, someone could thing the name is "Client Bluetooth Device Address", and there is no such thing. What we have is "Bluetooth Device Address" and "client".

**Comment by poussa** [comment]
LGTM

**Pull request merged**